src/librbd/api/Trash.cc

```diff
@@ -529,12 +529,48 @@ int Trash<I>::purge(IoCtx& io_ctx, time_t expire_ts,
 
   NoOpProgressContext remove_pctx;
   uint64_t list_size = to_be_removed.size(), i = 0;
-  for (const auto &entry_id : to_be_removed) {
-    r = librbd::api::Trash<I>::remove(io_ctx, entry_id, true, remove_pctx);
-    if (r < 0) {
-      return r;
+  int remove_err = 1;
+  while (!to_be_removed.empty() && remove_err == 1) {
+    remove_err = 0;
+    for (auto it = to_be_removed.begin(); it != to_be_removed.end(); ) {
+      trash_image_info_t trash_info;
+      r = Trash<I>::get(io_ctx, *it, &trash_info);
+      if (r == -ENOENT) {
+        // likely RBD_TRASH_IMAGE_SOURCE_USER_PARENT image removed as a side
+        // effect of a preceeding remove (last child detach)
```

Comment:
Actually I think it was my overlooking in #27521, when I introduced `TRASH_IMAGE_SOURCE_USER_PARENT`, to add it to the `trash_list` result filter used here.

As the `TRASH_IMAGE_SOURCE_USER_PARENT` images are removed automatically I think the trash purge should not be interested in them at all.

May be this was the regression you actually stepped on?

So I think we just have to remove `TRASH_IMAGE_SOURCE_USER_PARENT` condition when filtering `trash_entries`, and then we don't need this ` Trash::get` check. (Skipping other errors still looks useful).

Reply:
I actually considered simply tweaking that lambda but decided against that.  What if when the last child goes away `TRASH_IMAGE_SOURCE_USER_PARENT` image fails to be removed for some reason? Then `rbd trash purge` would always skip it and we are back to users having to resort to `rbd trash rm` on individual images.

Reply:
Ok. It makes sense.